### ISSUE TITLE
[HOLD for payment 2025-01-07] [$250] Android - Expense - Submit button jumps when proceeding to confirmation page

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#53182]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.67-1
**Reproducible in staging?:** Y
**Reproducible in production?:** N
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:** Y
**Email or phone of affected tester (no customers):** applausetester+232803@applause.expensifail.com
**Issue reported by:** Applause Internal Team

## Action Performed:
1. Launch the app.
2. Go to FAB > Create expense > Manual.
3. Enter amount > Next.
4. With the keyboard up, select any user at the top.

## Expected Result:
Submit button will not jump when proceeding to confirmation page.

## Actual Result:
Submit button jumps when proceeding to confirmation page.

## Workaround:
Unknown

## Platforms:

- [x] Android: Standalone
- [x] Android: HybridApp
- [ ] Android: mWeb Chrome
- [ ] iOS: Standalone
- [ ] iOS: HybridApp
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/94bed365-e0db-4648-b60a-0b298876e8a4
  
</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021861729998144839111</li>
        <li>Upwork Job ID: 1861729998144839111</li>
        <li>Last Price Increase: 2024-11-27</li>
<li>Automatic offers: </li>
<ul>
<li>suneox | Reviewer | 105169308</li>
<li>huult | Contributor | 105169309</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @sonialiap / @MitchExpensify</details>

**Comment by techievivek** [comment]
Not deemed to be a blocker, demoting it and sending to to external

**Comment by mountiny** [comment]
cc @kirillzyusko @c3024 This is something to look into/ follow up on later

**Comment by huult** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Submit button jumps when proceeding to confirmation page

### What is the root cause of that problem?

The issue occurs because the keyboard visibility affects the layout height when transitioning between screens.
Without dismissing the keyboard, the layout adjusts briefly to the keyboard's height before resizing to the new screen.

### What changes do you think we should make in order to solve the problem?

To resovle this issue we dismissing the keyboard before navigation ensures a stable layout transition, preventing the temporary jump.

```diff
// src/pages/iou/request/step/IOURequestStepParticipants.tsx#L110
    const goToNextStep = useCallback(() => {
+        Keyboard.dismiss();
        ...
    }
```


<details>
<summary>POC</summary>


</details>

### What alternative solutions did you explore? (Optional)

**Comment by huult** [comment]
## [Proposal updated](https://github.com/Expensify/App/issues/53182#issuecomment-2504065057)

- Updated new RCA
- Add alternative solutions

**Comment by suneox** [comment]
> ## [Proposal updated](https://github.com/Expensify/App/issues/53182#issuecomment-2504065057)
> * Updated new RCA
> * Add alternative solutions

@huult Your main solution doesn’t seem to work consistently

https://github.com/user-attachments/assets/5eddb9c7-c912-4286-9e1c-000199954292

Could you please provide more details about the RCA for your alternative solution?

**Comment by ikevin127** [comment]
# Proposal

### Please re-state the problem that we are trying to solve in this issue

Submit button jumps when proceeding to confirmation page.

### What is the root cause of that problem?

The issue seems to be caused by the recent implementation of android specific `KeyboardAvoidingView` component:

https://github.com/Expensify/App/blob/3cc88f5a83a979c5e588e01065083a3d714e69da/src/components/KeyboardAvoidingView/index.android.tsx#L61

from this PR:
- https://github.com/Expensify/App/pull/52392

which make it such that `ScreenWrapper` props like `shouldDismissKeyboardBeforeClose` (true by default) and `keyboardAvoidingViewBehavior` to have custom implementation and not work as expected anymore.

The same PR brought about changes related to how the keyboard / layout interactions behave and are being handled / animated / transitioned on **Android: Native**, which I think is also related to causing our glitch / delay when transitioning from one page to another where the keyboard is not needed and we have a button which is supposed to be fixed at the bottom.

### What changes do you think we should make in order to solve the problem?

> [!note]
> I also tried to play around with props like `shouldEnableKeyboardAvoidingView` and use `Keyboard.dismiss()` but none of those seem to fix the issue given the new android specific `KeyboardAvoidingView` implementation detailed above.

**TLDR**: Prevent displaying the `Submit {amount}` button until `Keyboard.isVisible` becomes `false`.

To do this we need to apply some changes to the [`FixedFooter`](https://github.com/Expensify/App/blob/3cc88f5a83a979c5e588e01065083a3d714e69da/src/components/FixedFooter.tsx#L15) component. We can do this in at least **two different ways**:
1. Add a new prop like `onlyShowWhenKeyboardIsHidden` and in `BaseSelectionList` where `FixedFooter` is used -> pass `onlyShowWhenKeyboardIsHidden={Platform.OS === CONST.PLATFORM.ANDROID}`.
2. Split the `FixedFooter` component into `index.tsx` which will remain unchanged and `index.android.tsx` where we will add specific logic to prevent displaying the `children` until `Keyboard.isVisible` becomes `false` (I know this is the recommended way).

For this proposal's sake I'm going with the first option (if selected we can further discuss the best way):
1. Modify `FixedFooter` as below, making sure to import `Keyboard` from `react-native` and the new prop type (boolean):

```ts
function FixedFooter({style, children, onlyShowWhenKeyboardIsHidden = false}: FixedFooterProps) {
    const styles = useThemeStyles();
    const isKeyboardVisible = Keyboard.isVisible();
    const shouldHide = onlyShowWhenKeyboardIsHidden ? isKeyboardVisible : false;

    if (!children || shouldHide) {
        return null;
    }

    return <View style={[styles.ph5, styles.pb5, styles.flexShrink0, style]}>{children}</View>;
}
```

2. In `BaseSelectionList` on [this line](https://github.com/Expensify/App/blob/3cc88f5a83a979c5e588e01065083a3d714e69da/src/components/SelectionList/BaseSelectionList.tsx#L852), pass `onlyShowWhenKeyboardIsHidden={Platform.OS === CONST.PLATFORM.ANDROID}`.

### Result video (before / after)

<details>
<summary>Android: Native (OnePlus 7T - Android 14 - Real device)</summary>

| Before | After |
| ------------- | ------------- |
| <video src="https://github.com/user-attachments/assets/e9ccb16d-e644-4669-acbd-d98b796cab9f"> | <video src="https://github.com/user-attachments/assets/00d72b54-424e-44d9-97f0-08e161a12d78"> |

</details>

**Comment by huult** [comment]
> Could you please provide more details about the RCA for your alternative solution?

Thanks, @suneox, for the review. I will provide more details tomorrow

**Comment by kirillzyusko** [comment]
We should just use `await KeyboardUtils.dismiss()` before calling a callback that would trigger a navigation and it should fix a problem.

A similar fix already used in `FormProvider` component

**Comment by kirillzyusko** [comment]
> const isKeyboardVisible = Keyboard.isVisible();

`Keyboard.isVisible()` is not a reactive and will not trigger a re-render when keyboard changes its state. So even if it works now it can be broken at any time in the future.

**Comment by ikevin127** [comment]
@kirillzyusko Thanks for the tip, wasn't aware of that util function!

## Updated [proposal](https://github.com/Expensify/App/issues/53182#issuecomment-2509522822)
- added improved alternative solution mentioned in https://github.com/Expensify/App/issues/53182#issuecomment-2511487903

**Comment by huult** [comment]
## [Proposal updated](https://github.com/Expensify/App/issues/53182#issuecomment-2504065057)

- Updated keyboard dismiss functionality to a utility function.

@suneox I've updated the proposal. Could you please check?

**Comment by ikevin127** [comment]
> I updated the proposal before the other proposal.

@suneox As per our guidelines, a proposal update only counts after posting the **updated proposal** message, which I did first 🙂

**Comment by suneox** [comment]
Thanks for all the updates. I’ll review them within a few hours.

**Comment by suneox** [comment]
Both proposals from @ikevin127 and @huult also use the suggestion provided in [this comment](https://github.com/Expensify/App/issues/53182#issuecomment-2511487903). However, @huult [proposal](https://github.com/Expensify/App/issues/53182#issuecomment-2504065057) appears safer as it does not break the lint checks, whereas @ikevin127 [proposal](https://github.com/Expensify/App/issues/53182#issuecomment-2509522822) just only apply suggestion change & doesn't address fixing lint checks. 

Regarding the RCA, @ikevin127 proposal correctly identifies the offending PR, while @huult proposal also identifies the issue occurring within `KeyboardAvoidingView`.



Considering the above, I think we can proceed with @huult [solution](https://github.com/Expensify/App/issues/53182#issuecomment-2504065057) because:

1. His proposal does not break the lint checks.
2. The update was made before @ikevin127 proposal, despite the delayed notification.

🎀 👀 🎀 C+ reviewed

**Comment by ikevin127** [comment]
Wow, didn't expect the `lint` to be an argument for why the proposal that was updated first according to the guidelines (mine) was not selected, especially when I suggested that the implementation should follow already existing `FormProvider` implementation pattern - which if reviewed properly would be clear that it doesn't break lint as it uses `.then` instead of `await`.

@techievivek Care to take a second look at the events timeline here ? 

I feel like the argument used for selecting the other proposal doesn't hold any real weight in this case and the guidelines were not followed by the reviewer.

Thank you!

**Comment by suneox** [comment]
> I feel like the argument used for selecting the other proposal doesn't hold any real weight in this case and the guidelines were not followed by the reviewer.

@ikevin127 Sorry if I missed something in our C+ [Process Doc](https://docs.google.com/document/d/1BvohU05MTaHnjOD_vwJv_aqDAirv-ChkyRnKCAvOVyQ) regarding notify updated proposal timeline. I couldn’t find any related documentation for this, nor could I locate it in the Contributing Guides.

https://github.com/Expensify/App/blob/b8eb7053094d419512adf9c1020e67b83356e671/contributingGuides/CONTRIBUTING.md?plain=1#L146-L153 

> which if reviewed properly would be clear that it doesn't break lint as it uses .then instead of await.

When we use async/await in this case we have to disable @lwc/lwc/no-async-await

![Screenshot 2024-12-03 at 01 37 31](https://github.com/user-attachments/assets/6cff1f65-3ef4-4deb-9dc6-41d2e0750fb2)

**Comment by huult** [comment]
Thank you, @suneox, for the review.

@techievivek, I am waiting for your assignment. Thank you!

**Comment by techievivek** [comment]
I think Kevin is talking about this point https://github.com/Expensify/App/blob/b8eb7053094d419512adf9c1020e67b83356e671/contributingGuides/CONTRIBUTING.md?plain=1#L149

```
please go back and edit your original proposal, then post a new comment to the issue in this format to alert everyone that it has been updated:
```

While I agree that, ideally, @huult should have posted a comment stating that the proposal was updated, they were already actively communicating and providing regular updates. Since they were the first to make the updates, we can proceed with their proposal.

**Comment by huult** [comment]
Thank you all, I will create a pull request within 24 hours.

**Comment by sonialiap** [comment]
I'm OOO Dec 16-20, adding a leave buddy

next steps:
- waiting for PR to be finished

**Comment by MitchExpensify** [comment]
@huult what is the latest on [the PR](https://github.com/Expensify/App/pull/53529)?

**Comment by suneox** [comment]
> @huult what is the latest on [the PR](https://github.com/Expensify/App/pull/53529)? 

I'm still reviewing the PR and will finish testing it today.

**Comment by MitchExpensify** [comment]
*Payment summary*:

- $250 @suneox requires payment [automatic offer](https://www.upwork.com/nx/wm/offer/105169308) (Reviewer)
- $250 @huult requires payment [automatic offer](https://www.upwork.com/nx/wm/offer/105169309) (Contributor)

**Comment by MitchExpensify** [comment]
# BugZero Checklist:

- [ ] **[Contributor]** Classify the bug:

<details>
<summary>Bug classification</summary>


Source of bug:
  - [ ] 1a. Result of the original design (eg. a case wasn't considered)
  - [ ] 1b. Mistake during implementation
  - [ ] 1c. Backend bug
  - [ ] 1z. Other:

Where bug was reported:
  - [ ] 2a. Reported on production (eg. bug slipped through the normal regression and PR testing process on staging)
  - [ ] 2b. Reported on staging (eg. found during regression or PR testing)
  - [ ] 2d. Reported on a PR
  - [ ] 2z. Other:

Who reported the bug:
  - [ ] 3a. Expensify user
  - [ ] 3b. Expensify employee
  - [ ] 3c. Contributor
  - [ ] 3d. QA
  - [ ] 3z. Other:

</details>

- [ ] **[Contributor]** The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake.

    Link to comment:

- [ ] **[Contributor]** If the regression was CRITICAL (e.g. interrupts a core flow) A discussion in [#expensify-open-source](https://app.slack.com/client/E047TPA624F/C01GTK53T8Q) has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner.

    Link to discussion:

- [ ] **[Contributor]** If it was decided to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps using the template below to ensure the same bug will not reach production again.

<details>
<summary>Regression Test Proposal Template</summary>


- [ ] **[BugZero Assignee]** Create a GH issue for creating/updating the regression test once above steps have been agreed upon.

    Link to issue:

## Regression Test Proposal
### Precondition:


-

### Test:


1.

Do we agree 👍 or 👎


</details>

**Comment by sonialiap** [comment]
@suneox please complete the [checklist](https://github.com/Expensify/App/issues/53182#issuecomment-2566726359). Once done we'll issue the payment :)

**Comment by huult** [comment]
@sonialiap I’ve quickly created a pull request to fix the regression test without needing to pick or rent out anything else. Could you please consider issuing a refund? Thank you.
